### PR TITLE
Command Palette on notebook. 

### PR DESCRIPTION
**Pull request by Carreau**
Mainly openig PR to start discussion. 

I would like to start working on a **command palette** (cf atom screenshot) for the notebook. Something aka Mac OS Spotlight. It's called **command palette** in atom, so let's use this name, we can bikeshed on the name later. 

This would allow to expose a much bigger number of action to end-user without binding to a keyboard shortcut, and made them easy to find.

From a prototype I did at SciPy, I think it can be done correctly in about a week. 
And by correctly I mean with respect to the current API we have for the notebook. We all know that current API is not perfect, but we have enough hook to make something useful in a short amount of time.

I think it would be nice small project with high visibility and rapid user feedback that might help a lot of user and push for 4.1 adoption. Also one of the things I enjoyed when joining the project was to get small things that help people a lot. We might also want some design work to be done on the UI. 
So I think it would be a nice project for @svurens, @jdfreder and me to collaborate on. 

Technically I am torn in between twitter [typeahead](https://github.com/twitter/typeahead.js) and jQuery [typeahead](http://www.runningcoder.org/jquerytypeahead/), twitter one have stalled since April, so I went for second one. So I have no issue if we change. 

Thought ? Would @svurens like to work on that ?

--- 

screen shot 2015-08-05 at 14 23 15

### CODE REVIEW

**Comment by Carreau** [comment]
Ok, I have a rough prototype that show something. 


screen shot 2015-08-05 at 15 28 33

**Comment by ellisonbg** [comment]
@Carreau thanks for posting this - I am *really* excited about this. We have needed this for a long time and I think it is a great idea of have @svurens start to work on this with you and @jdfreder.

**Comment by ellisonbg** [comment]
Would type ahead help the UI side of this:

https://twitter.github.io/typeahead.js/

**Comment by jdfreder** [comment]
@svurens and I are fixing some problems with his Canopy+Anaconda+standalone IPython install...

**Comment by Carreau** [comment]
> Technically I am torn in between twitter typeahead and jQuery typeahead, twitter one have stalled since April, so I went for second one. So I have no issue if we change.

> Would type ahead help the UI side of this:

@ellisonbg : as I wrote in initial description Twitter-typeahead have 0 commits since april, no responses to PRs and issues. So I'm slightly negative on it. 

Also the API seem to change dramatically in between patch version so i"m not convinced we should use it. So I'm investigating jQuery typeahead, which seem : 

  - better documented 
  - have more real-world example. 
  - maintained.

**Comment by jdfreder** [comment]
@Carreau , @svurens has his development install working, and he is going to open a PR against your branch for discussion : )

**Comment by Carreau** [comment]
@jdfreder, @svurens cleaned up a bit, appreciate review.

File: notebook/services/contents/filemanager.py [hunk 1]
```diff
@@ -45,7 +45,6 @@ def _post_save_script(model, os_path, contents_manager, **kwargs):
     log = contents_manager.log
 
     base, ext = os.path.splitext(os_path)
-    py_fname = base + '.py'
```
Comment (Carreau) [inline notebook/services/contents/filemanager.py]: This was unused, so I removed it.

File: notebook/static/notebook/js/quickhelp.js [hunk 2]
```diff
@@ -168,10 +184,8 @@ define([
         var element = $('<div/>');
 
         // The documentation
-        var doc = $('<div/>').addClass('alert alert-warning');
+        var doc = $('<div/>').addClass('alert alert-info');
         doc.append(
-            $('<button/>').addClass('close').attr('data-dismiss','alert').html('&times;')
-        ).append(
```
Comment (Carreau) [inline notebook/static/notebook/js/quickhelp.js]: This made no sens as there are no persistence of the hide this thing. So removed as well.

File: notebook/static/notebook/js/quickhelp.js [hunk 3]
```diff
@@ -235,41 +249,21 @@ define([
 
     QuickHelp.prototype.build_command_help = function () {
         var command_shortcuts = this.keyboard_manager.command_shortcuts.help();
-        return build_div('<h4>Command Mode (press <code>Esc</code> to enable)</h4>', command_shortcuts);
-    };
-
-    var special_case = { pageup: "PageUp", pagedown: "Page Down", 'minus': '-' };
```
Comment (Carreau) [inline notebook/static/notebook/js/quickhelp.js]: this is now baked into `humanize_sequence`.  The logic was convoluted and wrong.

File: notebook/static/notebook/js/notebook.js [hunk 4]
```diff
@@ -170,7 +171,9 @@ define(function (require) {
         // i) provide a margin between the last cell and the end of the notebook
         // ii) to prevent the div from scrolling up when the last cell is being
         // edited, but is too low on the page, which browsers will do automatically.
-        var end_space = $('<div/>').addClass('end_space');
+        var end_space = $('<div/>')
+            .addClass('end_space')
+            .html('<kbd>Cmd-Shift-P</kbd> for Command palette');
```
Comment (Carreau) [inline notebook/static/notebook/js/notebook.js]: What do you think of having a Tips thing here ? (I should probably remove before merge)

File: notebook/services/contents/filemanager.py (hunk 1 above)
Comment (jdfreder) [inline notebook/services/contents/filemanager.py]: :+1:

File: notebook/static/notebook/js/actions.js [hunk 5]
```diff
@@ -329,10 +329,19 @@ define(function(require){
             handler : function (env) {
                 env.pager.collapse();
             }
+        },
+        'command-palette': {
+            help_index : 'aa',
```
Comment (jdfreder) [inline notebook/static/notebook/js/actions.js]: index aa?

File: notebook/static/notebook/js/commandpalette.js [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+      shortcut = notebook.keyboard_manager.edit_shortcuts.get_shortcut_for_action_name(name);
+      if (shortcut) {
+        return 'edit-shortcut';
+      }
+      return 'no-shortcut';
+    };
+
+    var CommandPalette = function(notebook) {
+        if(!notebook){
+          throw new Error("CommandPalette take a notebook non-null mandatory arguement");
```
Comment (jdfreder) [inline notebook/static/notebook/js/commandpalette.js]: s/take/takes/

File: notebook/static/notebook/js/commandpalette.js [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+          }
+          before_close.ok = true; // avoid double call.
+        };
+
+        var mod = $('<div/>').addClass('modal').append(
+          $('<div/>').addClass('modal-dialog')
+          .append(
+            $('<div/>').addClass('modal-content').append(
+              $('<div/>').addClass('modal-body')
+              .css('padding', '7px')
```
Comment (jdfreder) [inline notebook/static/notebook/js/commandpalette.js]: You use .css in a few places, you should give the outer most div a `command-palette` class and style through less.  That will make it easier for users to add custom styling too.

File: notebook/static/base/js/keyboard.js [hunk 8]
```diff
@@ -225,6 +225,17 @@ define([
         }
         return dct;
     };
+    
+    ShortcutManager.prototype.get_shortcut_for_action_name = function(name){
```
Comment (jdfreder) [inline notebook/static/base/js/keyboard.js]: Can this be `getShortcut` or `get` ?  So then when it's called it looks like this `shortcutManager.getShortcut('bob')` or `shortcutManager.get('bob')` .

Comment (jdfreder) [inline notebook/static/base/js/keyboard.js]: I guess the in this file we are using underscore separated words, so `get_shortcut` ... :/

File: notebook/static/base/less/variables.less [hunk 9]
```diff
@@ -15,6 +15,9 @@
 @border-radius-base: 2px;
 @border-radius-large: 3px;
 @grid-gutter-width: 0px;
+@kbd-color: #888;
+@kbd-bg: transparent;
```
Comment (jdfreder) [inline notebook/static/base/less/variables.less]: I'm guessing `kbd` stands for keyboard dialog.  Why not `@commandpalette-*` ?

File: notebook/static/base/js/keyboard.js (hunk 8 above)
Comment (jdfreder) [inline notebook/static/base/js/keyboard.js]: I prefer `get`

**Comment by jdfreder** [comment]
I left a couple of in line comments, but I'd also like to leave a general comment: you don't have many comments (no pun intended) in your code, you may want to add some more so it's easier to digest.

Other than that, it's looking good!  @svurens can you review this too?

File: notebook/static/notebook/js/actions.js (hunk 5 above)
Comment (Carreau) [inline notebook/static/notebook/js/actions.js]: sure, I think it's important and should be near the top. What woudl you use ?

File: notebook/static/base/less/variables.less (hunk 9 above)
Comment (Carreau) [inline notebook/static/base/less/variables.less]: no, that'a a bootstrap variable.

File: notebook/static/base/js/keyboard.js (hunk 8 above)
Comment (Carreau) [inline notebook/static/base/js/keyboard.js]: the problem with get, is that you could get many things, action, action name, shortcut, it was get and i keept getting confuse for what I was getting for what.

**Comment by Carreau** [comment]
Added comment. even if I'm not sure it make things clearer, there is not a lot to explain.

File: notebook/static/base/js/keyboard.js (hunk 8 above)
Comment (svurens) [inline notebook/static/base/js/keyboard.js]: I think get_shortcut or get_action_shortcut is about the shortest you can get without making this function's name too general.

Comment (Carreau) [inline notebook/static/base/js/keyboard.js]: fine with me, renamed !

**Comment by Carreau** [comment]
Any extra-comments ? @jdfreder I suppose you woudl prefer your commonJS PR to be merged first ?

**Comment by jdfreder** [comment]
@Carreau no, no more comments.  It's okay, I can deal with the rebase on my end : )  I think I came along after you..

**Comment by jdfreder** [comment]
Thanks!  People will love this!

**Comment by damianavila** [comment]
Thanks! People will love this!

>I don't see any docs reference here... this is something to add later in another PR?

**Pull request merged**